### ISSUE TITLE
Read-only / Disabled widgets

### ISSUE THREAD

**Issue by heroboy** [ocornut/imgui#211]

**Comment by andoma** [comment]
+1 on this, currently i just decrease alpha to make it look disabled but widgets still respond to events (obviously). Would be nice to have this built-in

**Comment by ocornut** [comment]
Which widget in particular are you using it for?

**Comment by andoma** [comment]
Typically sliders and buttons

![image](https://cloud.githubusercontent.com/assets/216384/9627070/0fd17942-5162-11e5-92d5-ceac84795e7b.png)

vs

![image](https://cloud.githubusercontent.com/assets/216384/9627075/1ca260f0-5162-11e5-8504-7210ff4027dd.png)

(this is with adjusting global alpha)

**Comment by ocornut** [comment]
Probably rather low-priority.
There's two aspects to it - the interaction and the rendering. 
The interaction is probably easy to sort out. Outside api to be based on PushButtonRepeat()/PopButtonRepeat() (storing a bool + a stack) and check it in the right spot.

The rendering side is more work because we ought to decide on the design here, how are disabled state represented in general, how it can be part of the theme in a sensible lightweight manner without adding dozens of new entries into the theme system that the user would have to manipulate. Perhaps the theme system can compute certain colors based on HSV space operations and store that so widgets code don't have to do those calculation. e.g. we'd have a Colors[] array and a ColorsDisabled[] array that's auto computed. This tie somehow with some of the ground toward a potential new design (#184).

For now I suppose we could add the feature as affecting interaction only, probably wouldn't hurt but it would have to be documented as "visuals will change in a future version".

**Comment by andoma** [comment]
Just the interaction-parts is perfectly sufficient for me.

Btw, thanks for all your awesome work on this.

**Comment by Pagghiu** [comment]
In our software I've been handling the disabled state myself, "wrapping" existing imgui calls.

**Comment by ocornut** [comment]
Some notes. Are you looking for a "InteractionDisabled" or "EditDisabled/Readonly" flag ?

- Text box must still be selectable / copiable into clipboard.
- Scrollbar must be active
- How about tree nodes and collapsing header?
- Do we allow sliders and drags to be ctrl+clickable to be turned into a ? How about the hover feedback on drags (currently the background is highlighted).
- How about columns separators?
- How about buttons? If the flag is called ReadOnly vs Disabled how we expect that button works?

**Comment by heroboy** [comment]
I think "Disable" means disable all child widgets including the scrollbar, and preventing all mouse interactive.
And only a few widget has "Readonly" flag. Readonly widget can interact as normal, but the content will not be changed.
And now the readonly can easily implement in application level, for example:
```
float tmp = x;
ImGui::InputFloat("xxx",&x);
x = tmp;
```
I think this is good, because windows controls act like this.

**Comment by ocornut** [comment]
I think read-only should be global and inherited as well, so you can easily turn an entire block into read-only and still, scroll and copy data.

**Comment by benvanik** [comment]
@ocornut agreed - having something similar to button repeat would be nice. It'd also prevent the need for a bool disabled/bool read_only/etc on everything.

I was just running into this myself needing both read-only content and separately fully-disabled UI (no input/greyed out/etc). My hacky solution in lieu of a real one is to filter some of the mouse events in broad regions before handing them to imgui and to draw a giant semi-translucent rect over everything at the end. It's nasty :)

**Comment by ocornut** [comment]
Alright,

I have added a ImGuiInputTextFlags_ReadOnly flag to InputText() / InputTextMultiline() as an exception because I think it is a more common use case (to make blobs of text more naturally copiable, even though you can use the logging api for that too).

We need different flag for Disabled and global- ReadOnly. Earlier is easier to implement, later probably a bit more work, as we need to dig a little more inside widgets to make sure things like IsItemHovered() works but the widget doesn't react. So probably will add Disabled early on. It will have to be documented as "unfinished" because it won't affect the visuals just you but it WILL in the future.

**Comment by ocornut** [comment]
There's another usage pattern which I found useful in one app I made, which is to have widgets that *looks* disabled but are active. Here I have a set of variables that can be overriden. You can click the override checkbox but you can also directly click on disabled-lookin widget on the right-side and it automatically set the override flag as you click them. That's easily done with just changing text color. 

![disabled-override](https://cloud.githubusercontent.com/assets/8225057/9793505/aacba9f0-57dd-11e5-8f9c-e9358de3b242.PNG)

Probably won't affect this task here, but I wanted to share that it may be also a viable idea in some cases.

**Comment by ocornut** [comment]
@haddock07 in #330
> Is there a way to disable a widget (shown but with a disabled state)?
> For example, I would like to have a check box, in a disabled state, where it couldn't be activated or hovered, and displayed with a darker color. Of course, the function ImGui::Checkbox would always return false.
> I did a workaround by setting the same color for the 3 states (normal, hovered, active), but when the user clicks on the check box, the check mark is actually shown one frame before returning to the unchecked state.

**Comment by r-lyeh** [comment]
Hey, 

I was looking for possible CheckboxDisabled()/RadioDisabled()/MenuItemDisabled() widgets and found this thread instead. I guess I may share some feedback.

Rather than multiplexing the workload into infinite disabled widget variants that may be (very) time consuming, my suggestion would be rather to create a single BeginDisabled()/EndDisabled(); group that pushes disable state into the stack. From this point:

a) I would expect to be able to browse still the disabled group (scrollbars and tooltips would be active), but not able to interact with any variable modifier at all (checkboxes, selectables, combos, input boxes and so on).  Ie, all sub-widgets that may change the user variables shall not react to keyboard/mouse events (ever); any other will. So all my widget values are safe (constant) inside a disabled stack.

b) I would expect the rendered widgets would look similar to a blended widget+background. Achieved either by blending every sub-widget with alpha *0.5, or by doing fancy HSV to blend every rendered color with the background color.

Keep up the good work!

**Comment by teamimx** [comment]
+1 on the general idea of a disabled state :)

**Comment by leftspace89** [comment]
easy solution.
![GitHub Logo](http://image.prntscr.com/image/7a101af54ecd48569ae90942d062f6c2.png)

![GitHub Logo](http://image.prntscr.com/image/fcd0467475c24487b71316003a58386a.png)

![GitHub Logo](http://image.prntscr.com/image/d0a418e188f248b7ab057d6edf932f34.png)

**Comment by ratzlaff** [comment]
By chance, could you make a patch instead of a series of images?

**Comment by leftspace89** [comment]
thx :)

https://github.com/ocornut/imgui/compare/master...leftspace89:master

2017-02-06 17:48 GMT+03:00 Ben Ratzlaff <notifications@github.com>:

**Comment by sonoro1234** [comment]
+1 on the general idea of a disabled state :)

**Comment by qimmer** [comment]
+1 on the general idea of a disabled state

**Comment by codecat** [comment]
Any update on this?

**Comment by ldecarufel** [comment]
I'd just like to chime in and say that I also would like a way to disable groups of controls.

Maybe PushDisabledState()/PopDisabledState() functions would be the best, and follows ImGui's stack-based approach.

Maybe disabled controls could simply be drawn in the monochrome version of their regular color, multiplied by a predefined color in the styles.

**Comment by r-lyeh** [comment]
I just hacked it from ideas in this thread.

```
namespace ImGui {
    std::vector<float> alphas;
    void PushDisabled( bool disabled = true ) {
        alphas.push_back( GetStyle().Alpha );
        GetStyle().Alpha = disabled ? 0.25f : 1.0f;
    }
    void PopDisabled( int num = 1 ) {
        while( num-- ) {
            GetStyle().Alpha = alphas.back();
            alphas.pop_back();
        }
    }
    // Install: edit ImGui::IsMouseHoveringRect() and change return to:
    //   return rect_for_touch.Contains(g.IO.MousePos) && (GetStyle().Alpha >= 1);
}
```

![image](https://user-images.githubusercontent.com/878187/29967762-8f809376-8f18-11e7-9ae0-766b21397862.png)

**Comment by ocornut** [comment]
I have pushed the simplest form of disabled flag, but haven't exposed it (it's in `imgui_internal.h` for now)
It doesn't affect the style but you can apply an alpha multiplier fairly easily.
(Btw @r-lyeh why are you using your own stack and not `PushStyleVar` which does exactly that?)
```
    static bool disabled = false;
    ImGui::Checkbox("Disable", &disabled);
    if (disabled)
    {
        ImGui::PushItemFlag(ImGuiItemFlags_Disabled, true);
        ImGui::PushStyleVar(ImGuiStyleVar_Alpha, ImGui::GetStyle().Alpha * 0.5f);
    }
    [...]
    if (disabled)
    {
        ImGui::PopItemFlag();
        ImGui::PopStyleVar();
    }
```
![disabled](https://user-images.githubusercontent.com/8225057/31986165-b0c1aa56-b967-11e7-94e5-b192d7a3a57a.PNG)

**Comment by r-lyeh** [comment]
ah, i overlooked the api :)

**Comment by Cwiiis** [comment]
I like the solution using PushItemFlag, but it would be nice to be able to use it like you can style flags, so you could push a flag that causes items to be disabled before creating a window. If you push the item flags before creating a window now, you end up creating the debug window instead of affecting subsequent items.

**Comment by Cwiiis** [comment]
In my use-case, I have a UI where in its 'mobile' mode, there is one main view, then a series of smaller preview windows. When you press the preview windows, they become the main view (and the main view shrinks and joins the preview windows). One of these views is a control-panel and I disable controls in it so that you can't accidentally activate anything until it's a main view.

The small issue is that without being able to disable items across windows, I need to pass a flag around a few different functions when the code that creates these windows isn't really concerned with whether or not it's disabled.

It's a good point about what the behaviour of a disabled window should be though - maybe you shouldn't be able to move it? (this particular application disables all window movement anyway, but I could imagine that being useful)

**Comment by aerydna** [comment]
any update on when this feature  will be exposed ? it would be very useful. Anyway thanks for all the work you do!

**Comment by ianshmean** [comment]
Has this functionality made it into 1.70? I noticed a mention in the release notes of the gamepad functionality that was added in https://github.com/ocornut/imgui/commit/a26085ed53d07603a9fde0e164ad69773d119642 that was referenced above

**Comment by x1nixmzeng** [comment]
Bumping for interest.

My usage is a row of several buttons I want to show on the debug window at all times:

```cpp
if (ImGui::Button("Button 1")) { ... }
ImGui::SameLine();
if (ImGui::Button("Button 2")) { ... }
```

Exposing `ButtonEx` and `ImGuiButtonFlags_Disabled` does the trick, but remains part of imgui_internal and doesn't change the style.

**Comment by ocornut** [comment]
This is still only in the private api because it doesn't affect styling.
In the meanwhile you may use .e.g PushStyleVar(ImGuiStyleVar_Alpha, 0.6f) / PopStyleVar() to alter the transparency of those widgets.

**Comment by fr500** [comment]
This is still internal only right?
Asking because I use cimgui and I noticed I can't access PushItemFlag

**Comment by ocornut** [comment]
> This is still internal only right?

Right.

cimgui should now include all internals (in the SAME cimgui.h file instead of a separate cimgui_internal.h file which I believe is a critical design flaw).

**Comment by timbeaudet** [comment]
Bumping for more interest in this becoming exposed, specifically with RadioButtons at the moment but it would be nice to add the ReadOnly flag much like you can with text/input boxes (InputFloat) for instance.

**Comment by PazerOP** [comment]
Is this something that could potentially be resolved with a pull request? Or is it waiting on love that only ocornut can give?

**Comment by flamendless** [comment]
Any update on this?

> I have pushed the simplest form of disabled flag, but haven't exposed it (it's in `imgui_internal.h` for now)
> It doesn't affect the style but you can apply an alpha multiplier fairly easily.
> (Btw @r-lyeh why are you using your own stack and not `PushStyleVar` which does exactly that?)
> 
> ```
>     static bool disabled = false;
>     ImGui::Checkbox("Disable", &disabled);
>     if (disabled)
>     {
>         ImGui::PushItemFlag(ImGuiItemFlags_Disabled, true);
>         ImGui::PushStyleVar(ImGuiStyleVar_Alpha, ImGui::GetStyle().Alpha * 0.5f);
>     }
>     [...]
>     if (disabled)
>     {
>         ImGui::PopItemFlag();
>         ImGui::PopStyleVar();
>     }
> ```
> 
> ![disabled](https://user-images.githubusercontent.com/8225057/31986165-b0c1aa56-b967-11e7-94e5-b192d7a3a57a.PNG)

this doesnt work as the `PushItemFlag` and `PopItemFlag` are not exposed publicly

**Comment by bkaradzic** [comment]
This is my hack, compiled from ideas from here...

```cpp

namespace ImGui
{
	inline void PushEnabled(bool _enabled)
	{
		extern void PushItemFlag(int option, bool enabled);
		PushItemFlag(1<<2 /*ImGuiItemFlags_Disabled*/, !_enabled);
		PushStyleVar(ImGuiStyleVar_Alpha, ImGui::GetStyle().Alpha * (_enabled ? 1.0f : 0.5f) );
	}

	inline void PopEnabled()
	{
		extern void PopItemFlag();
		PopItemFlag();
		PopStyleVar();
	}
	
} // namespace ImGui
```

**Comment by ocornut** [comment]
```cpp
PushItemFlag(1<<2 /*ImGuiItemFlags_Disabled*/, !_enabled);
```
Please note we don't provide ABI compatibility and flags occasionally get shuffled or changed value. Best to use `imgui_internal.h` and you can restrict it to a single cpp file if helpful.

**Comment by melMass** [comment]
> ```c++
> PushItemFlag(1<<2 /*ImGuiItemFlags_Disabled*/, !_enabled);
> ```
> 
> Please note we don't provide ABI compatibility and flags occasionally get shuffled or changed value. Best to use `imgui_internal.h` and you can restrict it to a single cpp file if helpful.

I never wanted to open a specific issue for this, but it seems appropriate here, as I'm using it for this reason among others:

What are the reasons against using imgui_internal? I'm using it all over the place in custom widgets without too much issue updating it along with imgui versions.
Is this simply a warning because things a more prone to change than in imgui.h?

**Comment by ocornut** [comment]
> Is this simply a warning because things a more prone to change than in imgui.h?

@melMass Yes pretty much.

**Comment by ocornut** [comment]
Hello all,

While I don't have a 100% satisfying solution yet (as in my mind it ties closely to a rework of the styling system), this issue has been open for ridiculously long and to mitigate this I have at least added the `PushDisabled() / PopDisabled()` functions in imgui_internal.h (which potentially upgrading them to imgui.h in the future).

1) in the event you have function with same name in same namespace you'll get a link error
2) in the event you used custom functions I would like to encourage you to use those functions now and let me know if they are not a good fit for you.

Thanks!

```cpp
// PushDisabled()/PopDisabled()
// - Those are not yet exposed in imgui.h because we are unsure of how to alter the style in a way that works for everyone.
//   We may rework this. Hypothetically, a future styling system may set a flag which make widgets use different colors.
// - Feedback welcome at https://github.com/ocornut/imgui/issues/211
// - You may trivially implement your own variation of this if needed.
//   Here we test (CurrentItemFlags & ImGuiItemFlags_Disabled) to allow nested PushDisabled() calls.
void ImGui::PushDisabled()
{
    ImGuiContext& g = *GImGui;
    if ((g.CurrentItemFlags & ImGuiItemFlags_Disabled) == 0)
        PushStyleVar(ImGuiStyleVar_Alpha, g.Style.Alpha * 0.6f);
    PushItemFlag(ImGuiItemFlags_Disabled, true);
}

void ImGui::PopDisabled()
{
    ImGuiContext& g = *GImGui;
    PopItemFlag();
    if ((g.CurrentItemFlags & ImGuiItemFlags_Disabled) == 0)
        PopStyleVar();
}
```

**Comment by ocornut** [comment]
A few people have liked the message above about `PushDisabled()`/`PopDisabled()`. 
I would be interested in your feedback (whereas the Alpha multiply gives a satisfying look with your style) to see if we can promote this to `imgui.h` in the future.

**Comment by codecat** [comment]
Not sure if it has been suggested yet, but would a function `SetNextWidgetEnabled` make sense? Thinking about conditionally enabling (or disabling) certain widgets. Perhaps even passing the enabled/disabled state to `PushDisabled` (or `PushEnabled` then?) would be an improvement here:

```c++
ImGui::PushEnabled(canUseButton);
if (ImGui::Button("Do something")) {
  // ...
}
ImGui::PopEnabled();
```

Versus the double conditional test currently required:
```c++
if (!canUseButton) ImGui::PushDisabled();
if (ImGui::Button("Do something")) {
  // ...
}
if (!canUseButton) ImGui::PopDisabled();
```

**Comment by codecat** [comment]
As for the alpha, in my style it might be a little difficult to see that buttons are actually disabled:

![image](https://user-images.githubusercontent.com/136534/123093324-d8bed780-d42b-11eb-8db8-eb104fd2dad8.png)

But I'm sure I'd be able to make this more pronounced as mentioned in the code's comments, so this is not a huge deal to me. In your screenshots it looks great, imo.

**Comment by ocornut** [comment]
> Not sure if it has been suggested yet, but would a function SetNextWidgetEnabled make sense? 

It would make sense to have a `SetNextItemEnabled()`. 
Currently it's a little difficult to implement because we don't have standardized "footer" code for all widgets so I can't implement it asap.

>  Perhaps even passing the enabled/disabled state to PushDisabled (or PushEnabled then?) would be an improvement here:

Interesting. You could wrap it yourself with a `SetEnabled(bool v)` function I suppose?

> As for the alpha, in my style it might be a little difficult to see that buttons are actually disabled: [....]
> But I'm sure I'd be able to make this more pronounced as mentioned in the code's comments, 

Yes and no, my question ties to this specifically. We don't want you to have to modify this value in the code.
So I guess we should expose it in ImGuiStyle at the same time we expose `PushDisabled()` in imgui.h.

**Comment by ocornut** [comment]
> You could wrap it yourself with a SetEnabled(bool v) function I suppose?

Scrap that, a `SetEnabled(bool)` api would create too much confusion with the stack-based version. What if we did PushEnabled(false) / SetEnabled() ? 

I looked at the suggested boolean, `PushEnabled(bool)`,
It's a little difficult to apply alpha when nested:

```cpp
PushEnabled(false);
PushEnabled(true);
// What is value of Alpha at this point? how can we restore "full alpha"
```

```cpp
PushEnabled(false);
PushEnabled(true);
PushEnabled(false);
// Ditto
```

This would require use or store a "CurrentAlpha" (== `Alpha * (Disabled? DisabledAlpha : 1.0f)`)

We could decide that `PushEnabled(false) + PushEnabled(true)` would leave us disabled (aka Disabled spread to the whole stack) which becomes easier to implement but probably not obvious to user.

So I am tempted to try to figure out how to implement the SetNextItemXXX version first...

Without those changes, tentative commit (in `features/disabled` branch):
https://github.com/ocornut/imgui/commit/0fc170a9cf729c4d6d04af7a10f231edec8ed876

**Comment by Legulysse** [comment]
Hi !

For reference, the most satisfying "disable" mechanic I have used in production code is the BeginDisabledGroup(bool)/EndDisabledGroup from Unity :
https://docs.unity3d.com/ScriptReference/EditorGUI.BeginDisabledGroup.html
It brings the benefits pointed out by codecat above (PushEnabled/PopEnabled), in the sense that the bool parameter allows for more straightforward blocks of code by avoiding a lot of if() statements when surrounding widgets that may become disabled depending on some context.

When using nested calls, as soon as a "disable" state is pushed, everything down the line will stay disabled.
So in your example, PushEnabled(false) + PushEnabled(true) would indeed maintain the "disable" state.

The main benefit is that generally, when you want to disable a big chunk of interface, you don't want a random call inside to re-enable a field in the middle of it.
I saw this problem happening a lot when we used manual calls on the static GUI.enabled flag.
Swapping to Begin/End groups everywhere fixed a lot of stuff and made our tools much more stable.

Considering all this, if I had the choice between using a SetNextItemXXX and blocks of PushEnabled(bool)/PopEnabled, I would use the push/pop 100% of the time.

**Comment by ocornut** [comment]
Thanks for the feedback. I am reworking the API to take a bool, and you are right it makes more sense that one "disabled" in the stack should always be enough to enforce things (which incidentally is easier to implement, so fine by me!)

My only issue is that "bool disabled = true" goes the opposite way of `BeginMenu()`/`MenuItem()` using `bool enabled = true` BUT I've simultaneously been very eager to remove the bool parameters in the menu functions, so I am also looking at this now.

**Comment by Legulysse** [comment]
> My only issue is that `bool disabled = true` goes the opposite way of `BeginMenu()`/`MenuItem()` using `bool enabled = true` BUT I've simultaneously been very eager to remove the bool parameters in the menu functions, so I am also looking at this whole thing together.

I personally wouldn't mind the logic difference since the function name is very explicit on its behaviour and I always double check parameter names, even though consistency is important. If you plan to remove the menu items in the long run, having some inconsistency seems very acceptable.
As a side note, I consider menus and blocks of widgets as two very distinct contexts (menu items will probably have very granular disable conditions, widgets will probably be handled in groups), so having separate mechanisms doesn't feel like its a problem.

You could also go with PushEnabled(bool enabled = true) instead, but I believe it would be counter-intuitive when it comes to the stacking mechanism since the name would suggest a priority over the disabled state, whereas PushDisabled is way more explicit about the "disabled" state always having priority.

**Comment by ocornut** [comment]
Thanks! Was about to post my message and yours just arrived so this duplicate some of your thoughts.

----

For BeginMenu/MenuItem: the value for `enabled` is frequently computed inline at the call site and so it makes some sense that it stays an individual bool rather than named flag:
```cpp
MenuItem("Property", ..., has_selection);
MenuItem("Property", ..., selected_id == object_id);
```
Problem is a transition between enabled_is_true item logic to disabled_is_true block logic would be annoying.
So in theory it would be nice is the new API could use enabled_is_true logic but I mean this is super bizarre:

```cpp
BeginEnabledGroup(false)
```

I've spent 30 minutes browsing the internet trying to find a better noun or verb that would allow enable state use the true value (`enablement` ? `enabledness` ?!) but can't find a satisfying one....

> menu items will probably have very granular disable conditions, widgets will probably be handled in groups), so having separate mechanisms doesn't feel like its a problem.

I guess you are right about this.

FYI we also use a `_Disabled` flag for:
- `ImGuiSelectableFlags_Disabled`
- `ImGuiTableColumnFlags_Disabled`
- `ImGuiItemFlags_Disabled` // internal: this is what we're working on via a new Push/Begin api
- `ImGuiButtonFlags_Disabled` // internal

Both Selectable and Button ones are item override essentially the same thing we are dealing with, and they primarily exist because this new API we're working on didn't exist. If we had an hypothetical `SetNextItemEnabled(false)` API it would work everywhere and not require the existence of widget-specific `ImGuiSelectableFlags_Disabled`/`ImGuiButtonFlags_Disabled`.

Also torturing myself on whether to use `BeginDisabled()` vs `PushDisabled()` which got me into revaluating some of the wording choice in the API.

**Comment by Legulysse** [comment]
> FYI we also use a `_Disabled` flag ...

Then I guess it makes even more sense to use a "disabled" wording here, and I would argue that it is more in line with the overall logic of the mechanism : we intend to "disable" some widgets in a block, being enabled is an implicit default state.

If menu items are placed inside a disabled group, their enable state would mean something like "I could be enabled if the whole block around me was not disabled". It's a bit of an edge case that may be a bit bizarre, but I believe it's way less confusing than using the PushEnabled approach in 99% of the other use cases. 

> Also torturing myself on whether to use `BeginXXX()` vs `PushXXX()` which got me into revaluating some of the wording choice in the API.

Im kinda new to imgui so if most of the code uses the Begin prefix you should take this route, unity uses the same syntax for this mechanism so it's totally legit.

And btw, thank you for all your work on imgui !

**Comment by ocornut** [comment]
Added bool parameter to `PushDisabled(bool disabled = true)` with cc40ae210.

As it turns out: BeginXXX functions have bool return values, PushXXX functions don't.

Before moving this to public API I would like to investigate a few ambiguous things:
- [ ] double-check behavior of `IsItemHovered()` on disabled items (we have a `ImGuiHoveredFlags_AllowWhenDisabled` flag which should work but probably need some double-checking, also see #1940 and #3419
- [ ] behavior of click/dragging on disabled items and how it impact parent window

For the later https://github.com/rokups/imgui/commit/c24b470 was an attempt at reworking that, which I didn't merge then. But the "prevents window from being dragged when hovering a disabled item" property of that commit was implemented by a876ad877 which weirdly set HoveredIdDisabled without HoveredId.

I think to honor #3419 and set HoveredId for consistency we will end up more or less adopting https://github.com/rokups/imgui/commit/c24b470

**Comment by thedmd** [comment]
To control widgets enable/disable I made `ImScopeEnable` and used in like in example below. Just my 2 cents.
Maybe somebody will find it useful. : )

```cpp
struct ImScopeEnable
{
    ImScopeEnable(bool isEnabled);
    ~ImScopeEnable();

    bool IsEnabled() const;

    void Pop();

private:
    bool m_IsEnabled = false;
    bool m_IsReleased = false;
};

void ShowPanel(bool enable)
{
    // parent, when panel is disabled, all underlying widgets are disabled
    ImScopeEnable panelEnable(enable);

    // explicit management
    ImScopeEnable addButtonEnable(CanAddItems());
    if (ImGui::Button("Add")) ...;
    addButtonEnable.Pop();

    // or utilize scope to clean up state
    {
        ImScopeEnable removeButtonEnable(CanCremoveItems());
        if (ImGui::Button("Remove")) ...;
    }
}
```

And implementation if for curious souls. As usual, customization require `imgui_internal.h`:
```cpp
# include <imgui.h>
# include <imgui_internal.h>

ImScopeEnable::ImScopeEnable(bool isEnabled)
    : m_IsEnabled(isEnabled)
{
    auto parentIsEnabled = !(ImGui::GetCurrentWindow()->DC.ItemFlags & ImGuiItemFlags_Disabled);
    if (!parentIsEnabled)
        m_IsEnabled  = false;

    if (!isEnabled && parentIsEnabled)
    {
        m_IsReleased = false;
        ImGui::PushItemFlag(ImGuiItemFlags_Disabled, true);
        ImGui::PushStyleVar(ImGuiStyleVar_Alpha, 0.5f);
    }
    else
        m_IsReleased = true;
}

ImScopeEnable::~ImScopeEnable()
{
    Pop();
}

bool ImScopeEnable::IsEnabled() const
{
    return m_IsEnabled;
}

void ImScopeEnable::Pop()
{
    if (m_IsReleased)
        return;

    m_IsReleased = true;

    ImGui::PopStyleVar();
    ImGui::PopItemFlag();
}
```

**Comment by ocornut** [comment]
Thank you @thedmd!
With the new API I think ImScopeEnable can be simplified to 

```
ImScopeEnable(bool isEnabled)
{
  ImGui::PushDisabled(!isEnabled);
}
void ImScopeEnable::Pop()
{
    if (m_IsReleased)
        return;
    m_IsReleased = true;
    ImGui::PopDisabled();
}
```

The hardcoded alpha multiplier (0.6) will be exposed when we make the API public.

**Comment by ocornut** [comment]
Still a couple of things to work on but I've already pushed those commits

4a100f7 - Nav: Disabled items are not candidate for default focus. (#211, #787) + simplify handling of ImGuiButtonFlags_PressedOnDragDropHold path.
97638bf - BeginMenu: remove unnecessary call to ItemHoverable() which is already done by the Selectable() call.
f668240 - InputInt/InputFloat: When used with Steps values and _ReadOnly flag, the step button look disabled. (#211)
ba1c846 - Internals: Removed ImGuiButtonFlags_Disabled (which had inconsistent behavior) in favor of ImGuiItemFlags_Disabled. Selectable()'s ImGuiSelectableFlags_Disabled now uses the later. (#211)
2952525 - Disabled items more consistently release active id if the active item got disabled. (#211)
cac869b - Disabled: don't prevent Selectable() from being selected. (#211) + BeginMenu()/MenuItemEx() uses PushDisabled() the standard way.
5dd1e38 - Disabled: disabled items set HoveredId, allowing e.g. HoveredIdTimer to function. (#211, #3419) + Menus: fix hovering a disabled menu or menu item not closing other menus.

Lots of small inconsistencies and issues fixed, unless there's any surprise I think next in line is to add a demo and move the api to public space.

**Comment by ocornut** [comment]
Renamed to `BeginDisabled()` / `EndDisabled()` and pushed on master.
Added `style.DisabledAlpha` and `ImGuiStyleVar_DisabledAlpha`, defaulting to 0.60f.

Some more done since last updated
- 638e1a721be3f7f9a99ca2836165d6c12a6f0b4c -  Disabled: fixed IsItemHovered() returning true on disabled item when navigated to.
- 246ad6c03c419560c422f5edbb0b5727fb40b405 -  Disabled: items more consistently release active id if the active item gets disabled. 
- fb4bbeb033254e0139b820525b9338262a1e03e6 - Disabled: fixed IsItemHovered() if popped disabled state after item, or when using Selectable_Disabled.
+ 6b8a059fc99f27438010c92aae0b05903f27b7d1 and 90bf996e1a0abc1e798c175d403706464230c874 which were intermediary steps toward being able to implement `SetNextItemDisabled()`, but right one I haven't found a satisfying solution to implement this (it's not easy for variety of reasons). Will get back to it later.

**Comment by azchatlanin** [comment]
How to deal with it? If you make the button inactive when you press it!
```c++
static bool button_disable = false;
if (button_disable) ImGui::BeginDisabled();
if (ImGui::Button("OK", ImVec2(120, 0)))  { button_disable = true; }
if (button_disable) ImGui::EndDisabled();
```

**Comment by codecat** [comment]
I guess you would just have to make a temporary variable:
```c++
static bool button_disable = false;
bool was_disabled = button_disable;
if (was_disabled) ImGui::BeginDisabled();
if (ImGui::Button("OK", ImVec2(120, 0)))  { button_disable = true; }
if (was_disabled) ImGui::EndDisabled();
```

**Comment by Xeverous** [comment]
Hello, I'm reviving an old issue (not sure why it is still open) because I noticed a regression in my project after updating the library.

Git bisect told me fb4bbeb033254e0139b820525b9338262a1e03e6 is the first bad commit and this is the issue: I was intentionally following many disabled widgets with `if (ImGui::IsItemHovered()) { ... }` to showcase on-hover tooltips to explain why these elements were disabled (or any other related information).

![image](https://github.com/ocornut/imgui/assets/20820409/79d00ba5-66e6-43c8-b73b-20ed86939eac)

```cpp
if (ImGui::BeginMenu("Menu")) {
	ImGui::MenuItem("A disabled option", nullptr, false, false);

	if (ImGui::IsItemHovered()) {
		ImGui::BeginTooltip();
		ImGui::PushTextWrapPos(ImGui::GetFontSize() * 35.0f);
		ImGui::TextUnformatted("An explanation why it is disabled");
		ImGui::PopTextWrapPos();
		ImGui::EndTooltip();
	}

	ImGui::EndMenu();
}
```

How can I implement such behavior after the linked commit?

**Comment by cfillion** [comment]
`ImGuiHoveredFlags_AllowWhenDisabled` or more generally: `IsItemHovered(ImGuiHoveredFlags_ForTooltip)` (v1.90+). Since 1.89.7 there's also a helper function for the common IsItemHovered+BeginTooltip pattern:

```cpp
ImGui::MenuItem("A disabled option", nullptr, false, false);
if(ImGui::BeginItemTooltip()) {
  ImGui::PushTextWrapPos(ImGui::GetFontSize() * 35.0f);
  ImGui::TextUnformatted("An explanation why it is disabled");
  ImGui::PopTextWrapPos();
  ImGui::EndTooltip();
}
```

Since v1.89.4 the return value of `BeginTooltip` should be checked.

**Comment by ocornut** [comment]
Closing this, indeed there's no reason to keep this open. This is the 8th oldest open issue :)

FYI i don't yet have a good implementing for `SetNextItemDisabled()` (it's quite tricky for obscure reason), but I'll eventually will.
